### ISSUE TITLE
Smart assertions eagerly evaluate rhs of logical expressions

### ISSUE THREAD

**Issue by minedeljkovic** [zio/zio#6886]
Smart assertions eagerly evaluate rhs of logical expressions, even if it's not logically necessary, so such example assertion fails if `option` is empty:
```Scala
assertTrue(option.isEmpty || option.get == 1)

// fails with
// • Option was None
// option.isEmpty || option.get == 1
// option = None
```
while logically it shouldn't fail.

**Comment by adamgfraser** [comment]
I'm not sure if the issue is eagerly evaluating it or just not properly handling it once it has been evaluated. Evaluating the right hand side only if the left hand side was false could potentially be more efficient, but even if we evaluate both sides and have one side that succeeds and one side that fails we should then succeed if the `||` operator is used.

**Comment by minedeljkovic** [comment]
> if we evaluate both sides and have one side that succeeds and one side that fails we should then succeed if the || operator is used

Indeed, this is probably easier to implement too. And I'm not sure the aim here is extremely high efficiency, since it's test code.
